File: resetgo/internal/d4/d4.go
// Package d4 implements the IEEE 1284.4 (Dot4 / D4) protocol used to talk to
// Epson printers over a bidirectional transport, ported from the Python
// ez-reset tool.
package d4

import (
	"encoding/binary"
	"fmt"
	"log"
	"time"

	"ezreset/internal/control"
	"ezreset/internal/transport"
)

// Packet is a single D4 packet.
type Packet struct {
	PSID    int
	SSID    int
	Credit  int
	Control int
	Payload []byte
}

// Command is a D4 control command.
type Command int

const (
	CmdInit         Command = 0
	CmdOpenChannel  Command = 1
	CmdCloseChannel Command = 2
	CmdCredit       Command = 3
	CmdCreditReq    Command = 4
	CmdExit         Command = 8
	CmdGetSocketID  Command = 9
)

var errors = map[byte]string{
	0x80: "Malformed packet",
	0x81: "No credit",
	0x82: "Reply without command",
	0x83: "Packet too big",
	0x84: "Channel not open",
	0x85: "Unknown Result",
	0x86: "Credit overflow",
	0x87: "Bad command/reply",
}

// D4Error is a protocol-level error.
type D4Error struct {
	Msg string
}

func (e *D4Error) Error() string { return e.Msg }

// Channel is an open D4 socket.
type Channel struct {
	d4       *D4
	SSID     int
	PSID     int
	MTU      int
	txCredit int
	rxCredit int
	rxMax    int
	rxQueue  []Packet
}

// Open opens the channel and grants initial receive credits.
func (c *Channel) Open() error {
	c.d4.OpenChannel(c)
	c.d4.Credit(c, c.rxMax)
	c.rxCredit += c.rxMax
	return nil
}

// Close closes the channel.
func (c *Channel) Close() error {
	c.d4.CloseChannel(c)
	return nil
}

func (c *Channel) ensureCredit() {
	for c.txCredit < 1 {
		if granted, err := c.d4.CreditRequest(c, 0); err == nil && granted >= 1 {
			return
		}
		time.Sleep(100 * time.Millisecond)
	}
}

// Write sends data over the channel, splitting into MTU-sized frames.
func (c *Channel) Write(data []byte, progress func(int)) error {
	for len(data) > 0 {
		control := 0
		payload := data
		if len(data) > c.MTU-6 {
			payload = data[:c.MTU-6]
		}
		control |= 2
		data = data[len(payload):]

		credit := c.rxMax - c.rxCredit
		if credit > 0xFF {
			credit = 0xFF
		}
		packet := Packet{PSID: c.PSID, SSID: c.SSID, Credit: credit, Control: control, Payload: payload}
		c.rxCredit += credit

		c.ensureCredit()
		if err := c.d4.WritePacket(c, packet); err != nil {
			return err
		}

		if progress != nil {
			progress(len(payload))
		}
	}
	return nil
}

// Read returns the next received packet.
func (c *Channel) Read() (Packet, error) {
	credit := c.rxMax - c.rxCredit
	if credit > 0xFF {
		c.d4.Credit(c, credit)
		c.rxCredit += credit
	}
	return c.d4.ReadPacket(c)
}

// D4 is the top-level protocol handler.
type D4 struct {
	transport transport.Transport
	channels  map[int]*Channel
}

// New creates a D4 handler and enters 1284.4 mode on the transport.
func New(t transport.Transport) (*D4, error) {
	d := &D4{
		transport: t,
		channels:  map[int]*Channel{0x00: {d4: nil, SSID: 0x00, txCredit: 1}},
	}
	d.channels[0x00].d4 = d

	// Drain any queued periodic status messages.
	if err := t.Drain(); err != nil {
		return nil, err
	}

	// Escape other modes, enter 1284.4 mode.
	if err := t.Write([]byte("\x00\x00\x00\x1b\x01@EJL 1284.4\n@EJL\n@EJL\n")); err != nil {
		return nil, err
	}
	if _, err := t.Read(8); err != nil {
		return nil, err
	}

	d.channels[0x00].d4 = d
	if err := d.Init(); err != nil {
		return nil, err
	}
	return d, nil
}

func (d *D4) getFreePSID() (int, error) {
	for i := 0; i < 0x100; i++ {
		if _, ok := d.channels[i]; !ok {
			return i, nil
		}
	}
	return 0, &D4Error{Msg: "No free PSIDs to allocate to channel open."}
}

// WritePacket serializes and writes a packet to the transport.
func (d *D4) WritePacket(channel *Channel, packet Packet) error {
	length := 6 + len(packet.Payload)
	header := make([]byte, 6)
	binary.BigEndian.PutUint16(header[0:2], uint16(packet.PSID))
	binary.BigEndian.PutUint16(header[2:4], uint16(packet.SSID))
	binary.BigEndian.PutUint16(header[4:6], uint16(length))
	header[4] = byte(packet.Credit)
	header[5] = byte(packet.Control)

	data := append(header, packet.Payload...)
	log.Printf("> %x", data[:min(len(data), 0x100)])
	if err := d.transport.Write(data); err != nil {
		return err
	}
	channel.txCredit--
	return nil
}

// ReadPacket returns the next packet destined for the given channel.
func (d *D4) ReadPacket(channel *Channel) (Packet, error) {
	for len(channel.rxQueue) == 0 {
		if err := d.readNextPacket(); err != nil {
			return Packet{}, err
		}
	}
	p := channel.rxQueue[0]
	channel.rxQueue = channel.rxQueue[1:]
	return p, nil
}

func (d *D4) readNextPacket() error {
	header, err := d.transport.Read(6)
	if err != nil {
		return err
	}
	psid := int(header[0])
	ssid := int(header[1])
	length := int(binary.BigEndian.Uint16(header[2:4]))
	credit := int(header[4])
	control := int(header[5])
	log.Printf("< %x", header)

	payload, err := d.transport.Read(length - 6)
	if err != nil {
		return err
	}
	log.Printf("< %x", append(header, payload...))

	channel, ok := d.channels[psid]
	if !ok {
		log.Printf("Received packet for closed socket ID %d", psid)
		return nil
	}

	channel.txCredit += credit
	channel.rxCredit--
	channel.rxQueue = append(channel.rxQueue, Packet{
		PSID:    psid,
		SSID:    ssid,
		Credit:  credit,
		Control: control,
		Payload: payload,
	})
	return nil
}

// Command sends a D4 control command and returns the response payload.
func (d *D4) Command(command Command, payload []byte) ([]byte, error) {
	if command != CmdInit && command != CmdExit {
		if d.channels[0].txCredit < 1 {
			return nil, &D4Error{Msg: "no credit on control channel"}
		}
	}

	packet := Packet{PSID: 0x00, SSID: 0x00, Credit: 1, Control: 0x00, Payload: append([]byte{byte(command)}, payload...)}
	if err := d.WritePacket(d.channels[0], packet); err != nil {
		return nil, err
	}
	res, err := d.ReadPacket(d.channels[0])
	if err != nil {
		return nil, err
	}

	if res.PSID != 0 {
		return nil, &D4Error{Msg: fmt.Sprintf("unexpected PSID %d", res.PSID)}
	}

	if len(res.Payload) > 0 && res.Payload[0] == 0x7F {
		msg := errors[res.Payload[3]]
		if msg == "" {
			msg = fmt.Sprintf("0x%x", res.Payload[3])
		}
		log.Printf("D4 error: %s", msg)
	}

	if len(res.Payload) == 0 || res.Payload[0] != byte(command)|0x80 || res.Payload[1] != 0 {
		return nil, &D4Error{Msg: fmt.Sprintf("unexpected response %x", res.Payload)}
	}

	return res.Payload[2:], nil
}

// Init performs the D4 initialization handshake.
func (d *D4) Init() error {
	resp, err := d.Command(CmdInit, []byte{0x10})
	if err != nil {
		return err
	}
	if len(resp) != 1 || resp[0] != 0x10 {
		return &D4Error{Msg: fmt.Sprintf("Init: unexpected response %x", resp)}
	}
	return nil
}

// Exit leaves D4 mode.
func (d *D4) Exit() error {
	_, err := d.Command(CmdExit, nil)
	return err
}

// GetSocketID resolves a named socket to its SSID.
func (d *D4) GetSocketID(name string) (int, error) {
	resp, err := d.Command(CmdGetSocketID, []byte(name))
	if err != nil {
		return 0, err
	}
	return int(resp[0]), nil
}

// OpenChannel opens a channel for the given SSID.
func (d *D4) OpenChannel(channel *Channel) error {
	psid := channel.SSID
	req := make([]byte, 12)
	req[0] = byte(psid)
	req[1] = byte(channel.SSID)
	binary.BigEndian.PutUint16(req[2:4], 0xFFFF)
	binary.BigEndian.PutUint16(req[4:6], 0xFFFF)
	binary.BigEndian.PutUint16(req[6:8], 0x0000)
	binary.BigEndian.PutUint16(req[8:10], 0x0000)
	binary.BigEndian.PutUint16(req[10:12], 0x0000)

	res, err := d.Command(CmdOpenChannel, req)
	if err != nil {
		return err
	}
	rPSID, rSSID, mtu, _, credit := parseOpenChannelResp(res)
	if rSSID != channel.SSID {
		return &D4Error{Msg: fmt.Sprintf("OpenChannel: SSID mismatch %d != %d", rSSID, channel.SSID)}
	}
	channel.PSID = rPSID
	channel.MTU = mtu
	channel.txCredit = credit
	d.channels[rPSID] = channel
	return nil
}

func parseOpenChannelResp(res []byte) (psid, ssid, mtu, maxCredit, credit int) {
	psid = int(res[0])
	ssid = int(res[1])
	mtu = int(binary.BigEndian.Uint16(res[2:4]))
	maxCredit = int(binary.BigEndian.Uint16(res[4:6]))
	credit = int(binary.BigEndian.Uint16(res[6:8]))
	return
}

// CloseChannel closes an open channel.
func (d *D4) CloseChannel(channel *Channel) error {
	req := []byte{byte(channel.PSID), byte(channel.SSID)}
	if _, err := d.Command(CmdCloseChannel, req); err != nil {
		return err
	}
	delete(d.channels, channel.PSID)
	return nil
}

// Credit grants receive credits to a channel.
func (d *D4) Credit(channel *Channel, amount int) error {
	req := make([]byte, 4)
	req[0] = byte(channel.PSID)
	req[1] = byte(channel.SSID)
	binary.BigEndian.PutUint16(req[2:4], uint16(amount))
	_, err := d.Command(CmdCredit, req)
	return err
}

// CreditRequest requests transmit credits from the printer.
func (d *D4) CreditRequest(channel *Channel, amount int) (int, error) {
	if amount == 0 {
		amount = 0xFFFF
	}
	req := make([]byte, 4)
	req[0] = byte(channel.PSID)
	req[1] = byte(channel.SSID)
	binary.BigEndian.PutUint16(req[2:4], uint16(amount))
	resp, err := d.Command(CmdCreditReq, req)
	if err != nil {
		return 0, err
	}
	_, _, granted := parseCreditReqResp(resp)
	d.channels[channel.PSID].txCredit += granted
	return granted, nil
}

func parseCreditReqResp(resp []byte) (psid, ssid, amount int) {
	psid = int(resp[0])
	ssid = int(resp[1])
	amount = int(binary.BigEndian.Uint16(resp[2:4]))
	return
}

// Channel opens (or returns) a named EPSON control channel.
func (d *D4) Channel(name string) (*Channel, error) {
	ssid, err := d.GetSocketID(name)
	if err != nil {
		return nil, err
	}
	return &Channel{d4: d, SSID: ssid, rxMax: 0x0001}, nil
}

// ControlBackend adapts D4 to the control.ControlBackend interface.
type ControlBackend struct {
	transport transport.Transport
	d4        *D4
	channel   *Channel
}

// NewControlBackend creates a D4-based control backend.
func NewControlBackend(t transport.Transport) *ControlBackend {
	return &ControlBackend{transport: t}
}

// Open enters D4 mode and opens the EPSON-CTRL channel.
func (b *ControlBackend) Open() error {
	d4, err := New(b.transport)
	if err != nil {
		return err
	}
	b.d4 = d4
	ch, err := d4.Channel("EPSON-CTRL")
	if err != nil {
		return err
	}
	if err := ch.Open(); err != nil {
		return err
	}
	b.channel = ch
	return nil
}

// Close closes the channel.
func (b *ControlBackend) Close() error {
	if b.channel != nil {
		_ = b.channel.Close()
	}
	return nil
}

// Send writes a command and reads the response payload.
func (b *ControlBackend) Send(command []byte) ([]byte, error) {
	if b.channel == nil {
		return nil, &control.BackendError{Msg: "Channel must be opened"}
	}
	if err := b.channel.Write(command, nil); err != nil {
		return nil, err
	}
	res, err := b.channel.Read()
	if err != nil {
		return nil, err
	}
	return res.Payload, nil
}

// Identify returns the IEEE 1284.4 device ID.
func (b *ControlBackend) Identify() (string, error) {
	return b.transport.Identify()
}

File: resetcplusplus/src/d4.cpp
#include "d4.h"

#include <chrono>
#include <thread>
#include <cstring>
#include <iostream>

namespace ezreset {

namespace {
const char* d4ErrorText(unsigned char code) {
    switch (code) {
        case 0x80: return "Malformed packet";
        case 0x81: return "No credit";
        case 0x82: return "Reply without command";
        case 0x83: return "Packet too big";
        case 0x84: return "Channel not open";
        case 0x85: return "Unknown Result";
        case 0x86: return "Credit overflow";
        case 0x87: return "Bad command/reply";
        default: return "Unknown D4 error";
    }
}
} // namespace

D4::D4(Transport* transport) : transport_(transport) {
    channels_[0x00] = new D4Channel(this, 0x00);
    channels_[0x00]->setTxCredit(1);

    transport_->drain();

    std::vector<unsigned char> enter = {
        '\x00','\x00','\x00','\x1b','\x01','@','E','J','L',' ','1','2','8','4','.','4',
        '\n','@','E','J','L','\n','@','E','J','L','\n'};
    transport_->write(enter);
    transport_->read(8);

    init();
}

D4::~D4() {
    for (auto& kv : channels_) delete kv.second;
}

void D4Channel::open() {
    d4_->openChannel(this);
    d4_->credit(this, rxMax_);
    rxCredit_ += rxMax_;
}

void D4Channel::close() {
    d4_->closeChannel(this);
}

void D4Channel::write(const std::vector<unsigned char>& data) {
    std::vector<unsigned char> remaining = data;
    while (!remaining.empty()) {
        int control = 0;
        std::vector<unsigned char> payload = remaining;
        if (static_cast<int>(remaining.size()) > mtu_ - 6) {
            payload.assign(remaining.begin(), remaining.begin() + (mtu_ - 6));
        }
        control |= 2;
        remaining.erase(remaining.begin(), remaining.begin() + payload.size());

        int credit = rxMax_ - rxCredit_;
        if (credit > 0xFF) credit = 0xFF;
        D4Packet packet{psid_, ssid_, credit, control, payload};
        rxCredit_ += credit;

        ensureCredit();
        d4_->writePacket(this, packet);
    }
}

void D4Channel::ensureCredit() {
    while (txCredit_ < 1) {
        if (d4_->creditRequest(this) >= 1) return;
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

D4Packet D4Channel::read() {
    int credit = rxMax_ - rxCredit_;
    if (credit > 0xFF) {
        d4_->credit(this, credit);
        rxCredit_ += credit;
    }
    return d4_->readPacket(this);
}

void D4::writePacket(D4Channel* channel, const D4Packet& packet) {
    int length = 6 + static_cast<int>(packet.payload.size());
    std::vector<unsigned char> header(6);
    header[0] = static_cast<unsigned char>(packet.psid);
    header[1] = static_cast<unsigned char>(packet.ssid);
    header[2] = static_cast<unsigned char>((length >> 8) & 0xFF);
    header[3] = static_cast<unsigned char>(length & 0xFF);
    header[4] = static_cast<unsigned char>(packet.credit);
    header[5] = static_cast<unsigned char>(packet.control);

    std::vector<unsigned char> data = header;
    data.insert(data.end(), packet.payload.begin(), packet.payload.end());
    transport_->write(data);

    channel->addTxCredit(-1);
}

void D4::readNextPacket() {
    std::vector<unsigned char> header = transport_->read(6);
    int psid = header[0];
    int ssid = header[1];
    int length = (header[2] << 8) | header[3];
    int credit = header[4];
    int control = header[5];
    std::vector<unsigned char> payload = transport_->read(length - 6);

    auto it = channels_.find(psid);
    if (it == channels_.end()) {
        std::cerr << "Received packet for closed socket ID " << psid << "\n";
        return;
    }
    D4Channel* channel = it->second;
    channel->addTxCredit(credit);
    channel->addRxCredit(-1);
    channel->pushQueue({psid, ssid, credit, control, payload});
}

D4Packet D4::readPacket(D4Channel* channel) {
    while (channel->queueEmpty()) {
        readNextPacket();
    }
    return channel->popQueue();
}

std::vector<unsigned char> D4::command(D4Command cmd,
                                       const std::vector<unsigned char>& payload) {
    if (cmd != D4Command::Init && cmd != D4Command::Exit) {
        if (channels_[0]->txCredit() < 1) {
            throw D4Error("no credit on control channel");
        }
    }
    std::vector<unsigned char> full;
    full.push_back(static_cast<unsigned char>(cmd));
    full.insert(full.end(), payload.begin(), payload.end());

    writePacket(channels_[0x00], {0x00, 0x00, 1, 0x00, full});
    D4Packet res = readPacket(channels_[0x00]);
    if (res.psid != 0) throw D4Error("unexpected PSID");

    if (!res.payload.empty() && res.payload[0] == 0x7F) {
        std::cerr << "D4 error: " << d4ErrorText(res.payload[3]) << "\n";
    }
    if (res.payload.empty() || res.payload[0] != (static_cast<int>(cmd) | 0x80)
        || res.payload[1] != 0) {
        throw D4Error("unexpected D4 response");
    }
    return std::vector<unsigned char>(res.payload.begin() + 2, res.payload.end());
}

void D4::init() {
    auto resp = command(D4Command::Init, {'\x10'});
    if (resp.size() != 1 || resp[0] != 0x10) throw D4Error("Init: unexpected response");
}

void D4::exit() {
    command(D4Command::Exit);
}

int D4::getSocketID(const std::string& name) {
    auto resp = command(D4Command::GetSocketID,
                        std::vector<unsigned char>(name.begin(), name.end()));
    return resp[0];
}

int D4::getFreePSID() {
    for (int i = 0; i < 0x100; i++) {
        if (channels_.find(i) == channels_.end()) return i;
    }
    throw D4Error("No free PSIDs to allocate to channel open.");
}

void D4::openChannel(D4Channel* channel) {
    int psid = channel->ssid();
    std::vector<unsigned char> req(12);
    req[0] = static_cast<unsigned char>(psid);
    req[1] = static_cast<unsigned char>(channel->ssid());
    req[2] = 0xFF; req[3] = 0xFF;
    req[4] = 0xFF; req[5] = 0xFF;
    req[10] = 0x00; req[11] = 0x00;

    auto res = command(D4Command::OpenChannel, req);
    int rPsid = res[0];
    int rSsid = res[1];
    int mtu = (res[2] << 8) | res[3];
    int credit = (res[6] << 8) | res[7];
    if (rSsid != channel->ssid()) throw D4Error("OpenChannel: SSID mismatch");
    channel->setPsid(rPsid);
    channel->setMtu(mtu);
    channel->setTxCredit(credit);
    channels_[rPsid] = channel;
}

void D4::closeChannel(D4Channel* channel) {
    std::vector<unsigned char> req = {
        static_cast<unsigned char>(channel->psid()),
        static_cast<unsigned char>(channel->ssid())};
    command(D4Command::CloseChannel, req);
    channels_.erase(channel->psid());
}

void D4::credit(D4Channel* channel, int amount) {
    std::vector<unsigned char> req = {
        static_cast<unsigned char>(channel->psid()),
        static_cast<unsigned char>(channel->ssid()),
        static_cast<unsigned char>((amount >> 8) & 0xFF),
        static_cast<unsigned char>(amount & 0xFF)};
    command(D4Command::Credit, req);
}

int D4::creditRequest(D4Channel* channel, int amount) {
    if (amount == 0) amount = 0xFFFF;
    std::vector<unsigned char> req = {
        static_cast<unsigned char>(channel->psid()),
        static_cast<unsigned char>(channel->ssid()),
        static_cast<unsigned char>((amount >> 8) & 0xFF),
        static_cast<unsigned char>(amount & 0xFF)};
    auto resp = command(D4Command::CreditRequest, req);
    int granted = (resp[2] << 8) | resp[3];
    channels_[channel->psid()]->addTxCredit(granted);
    return granted;
}

D4Channel* D4::channel(const std::string& name) {
    int ssid = getSocketID(name);
    return new D4Channel(this, ssid);
}

// ---- D4ControlBackend ----

D4ControlBackend::~D4ControlBackend() {
    close();
}

void D4ControlBackend::open() {
    d4_ = new D4(transport_);
    channel_ = d4_->channel("EPSON-CTRL");
    channel_->open();
}

void D4ControlBackend::close() {
    if (channel_) { channel_->close(); delete channel_; channel_ = nullptr; }
    if (d4_) { delete d4_; d4_ = nullptr; }
}

std::vector<unsigned char> D4ControlBackend::send(const std::vector<unsigned char>& command) {
    if (!channel_) throw std::runtime_error("Channel must be opened");
    channel_->write(command);
    D4Packet res = channel_->read();
    return res.payload;
}

std::string D4ControlBackend::identify() {
    return transport_->identify();
}

} // namespace ezreset
